File: R/gene_enrichment.R
#' exclude certain values from a table ,and convert to a vector
#'
#' @param data table of values
#' @param exclude matching table of booleans, where TRUE indicates values to exclude
#'
#' @export
#'
#' @return vector of values, with NA values removed
exclude_and_get_vector <- function(data, exclude) {
    data[exclude] = NA
    data = as.vector(data)
    
    return(data[!is.na(data)])
}

#' compute de novo enrichment for a gene
#'
#' @param de_novos table of de novos
#' @param n_male number of males
#' @param n_female number of females
#' @param symbol HGNC symbol for a gene
#' @param severity table of per base and per allele severity scores, for every gene
#' @param mu_rate table of mutation rates
#'
#' @export
#'
#' @return vector of p-value of de novo enrichment for a gene
compute_pvalue <- function(de_novos, n_male, n_female, symbol, severity, mu_rate) {
    
    # select the mutations for the given gene
    de_novos = de_novos[de_novos$gene == symbol, ]
    
    # count the samples in the cohort, so we can calculate expected mutations
    nsample = n_male + n_female
    if (de_novos$chr[1] == "X") {
        nsample = (n_male + n_female * 2) / 2
    }
    
    bases = c('A', 'C', 'G', 'T')
    
    severity = severity[severity$gene == symbol, ]
    positions = severity$pos
    ref_allele = severity$ref
    severity = severity[, bases]
    nsnv_o = sum(severity)
    mu_rate = mu_rate[mu_rate$gene == symbol, bases]
    
    # recode for LOF mutations
    severity[severity == 2] = 1
    severity[severity == 3] = 0
    
    if (nrow(severity) == 0) { return(NULL) }
    
    # downweight the severities by the rate across the alleles at each site
    severity = severity - apply((severity * mu_rate), 1, sum)
    
    # get the expected mutation rates per base per site for the gene
    mu_rate = nsample * 2 * mu_rate
    
    if (any(is.na(severity))) { warning(symbol, " has NA in severity") }
    
    # check if the position is in the CCDS (not included promoter region)
    de_novos = de_novos[de_novos$pos %in% positions, ]
    
    nsnv = nrow(severity)
    ndenovo = nrow(de_novos)
    scores = 0
    
    # if there aren't any sites with mutation rates, or we don't have any de
    # novos, give the gene a p-value of 1
    if (nrow(mu_rate) <= 0 | ndenovo == 0) {
        p_value = 1
        p_unweighted = 1
    } else {
        for (idx in seq(nrow(de_novos))) {
            if (de_novos$ref[idx] != ref_allele[positions == de_novos$pos[idx]]) {
                stop(de_novos$chr[idx], de_novos$pos[idx], de_novos$ref[idx], "do not match ref\n")
            }
            scores = scores + severity[positions == de_novos$pos[idx], de_novos$alt[idx]]
        }
        
        # exclude data from non-mutated, i.e. unchanged sites
        exclude = mu_rate == 0
        mu_rate = exclude_and_get_vector(mu_rate, exclude)
        severity = exclude_and_get_vector(severity, exclude)
        
        # compute p-values, using saddlepoint and standard poisson approaches
        p_value = double_saddle_point_approximation(scores, mu_rate, severity)
        p_unweighted = 1 - ppois(nrow(de_novos) - 1, sum(mu_rate))
    }
    
    values = list('symbol'=symbol, 'cohort_n'=nsample, 'nsnv_o'=nsnv_o,
        'n_sites'=nsnv, 'n_de_novos'=ndenovo, 'scores'=round(scores, 3),
        'p_value'=p_value, 'p_unweighted'=p_unweighted)
    
    return(values)
}

File: fitDNM/gene_enrichment.py

import math

from numpy import array
from scipy.stats import poisson

from fitDNM.saddlepoint import saddlepoint

def downweight_severity(data):
    ''' downweight severities by the rate across the alleles at each site
    
    severity = severity - (severity * mu_rate).apply(sum, axis=0)
    '''
    
    data = data.copy()
    data['value'] = data['prob'] * data['score']
    recode = dict([ (pos, sum(x['value'])) for pos, x in data.groupby('pos') ])
    data['score'] = data['score'] - data['pos'].map(recode)
    del data['value']
    
    return data['score']

def harmonise_data(rates, severity, symbol):
    ''' merge mutation rates and severity score datasets
    
    Args:
        rates: pandas DataFrame that includes a 'prob' column
        severity: pandas DataFrame that includes a 'score' column
        symbol: HGNC symbol for gene
    
    Returns:
        tuple of (mutation probabilities, severity score, position, )
    '''
    
    if len(rates) == 0 or len(severity) == 0:
        raise ValueError('empty rates or empty severity dataset!')
    
    data = rates.merge(severity, on=['gene', 'chrom', 'pos', 'ref', 'alt'])
    
    if len(data) == 0:
        raise ValueError('no shared sites between rates and severity datasets!')
    
    data = data[data['gene'] == symbol]
    
    assert len(set(data['chrom'])) == 1
    
    data = data[~data[['pos', 'alt']].duplicated()]
    return data.sort_values(['pos', 'alt'])

def get_expected_rates(data, male, female):
    ''' calculate expected number of mutations per site.
    
    Args:
        data: pandas DataFrame, which includes columns for 'chrom' (for
            chromosome) and 'prob' (for mutation probability)
        male: number of male probands
        female: number of female probands
    
    Returns:
        Mutation probabilies scaled so that each is the chance of observing a
        mutation at that site, given the number of male and female probands in
        the cohort.
    '''
    
    # count the samples in the cohort, so we can calculate expected mutations
    # TODO: add in a better chromX adjustment
    count = male + female
    if set(data['chrom']) == set(['X']):
        count = male / 2 + female
    
    # get the expected mutation rates per base per site for the gene
    return count * 2 * data['prob']

def enrichment(de_novos, n_male, n_female, symbol, severity, rates):
    ''' compute de novo enrichment for a gene
    
    Args:
        de_novos: table of de novos
        n_male: number of males
        n_female: number of females
        symbol: HGNC symbol for a gene
        severity: table of per base and per allele severity scores, for every gene
        rates: table of mutation rates
    
    Returns:
        dictionary of de novo enrichment results for a gene
    '''
    
    data = harmonise_data(rates, severity, symbol)
    
    data['score'] = downweight_severity(data)
    data['prob'] = get_expected_rates(data, n_male, n_female)
    
    if any(data['score'].isnull()):
        print("{} has NA in severity".format(symbol))
    
    # intersect the de novos with rate and severity data. This restricts
    # de novos to be within the coding sequence of the gene.
    data_prefixed = data.chrom.loc[0].startswith('chr')
    dnm_prefixed = de_novos.chrom.loc[0].startswith('chr')
    if data_prefixed and not dnm_prefixed:
        de_novos['chrom'] = 'chr' + de_novos['chrom']
    de_novos = de_novos.merge(data, on=['gene', 'chrom', 'pos', 'ref', 'alt'])
    observed_score = sum(de_novos['score'])
    
    p_value = saddlepoint(observed_score, data['prob'], data['score'])
    p_unweighted = poisson.sf(len(de_novos) - 1, sum(data['prob']))
    
    return {'symbol': symbol, 'gene_scores': sum(data['score']),
        'sites': len(set(data['pos'])), 'de_novos': len(de_novos),
        'de_novos_score': round(observed_score, 3), 'p_value': p_value,
        'p_unweighted': p_unweighted}
